{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import random\n",
    "# import sys\n",
    "# import warnings\n",
    "from pathlib import Path\n",
    "\n",
    "# import DataSets_contigs as ds\n",
    "# import kaplanmeier as km\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# import numpy as np\n",
    "# import optuna\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "import _config as cfg\n",
    "# import pymysql\n",
    "# from imblearn.over_sampling import SMOTE\n",
    "# from sklearn import svm\n",
    "# from sklearn.metrics import (\n",
    "#     accuracy_score,\n",
    "#     f1_score,\n",
    "#     precision_score,\n",
    "#     recall_score,\n",
    "#     roc_auc_score,\n",
    "# )\n",
    "# from sklearnex import patch_sklearn\n",
    "\n",
    "# patch_sklearn()\n",
    "# warnings.filterwarnings(\"ignore\")\n",
    "# random.seed(1024)\n",
    "# np.random.seed(1024)\n",
    "\n",
    "results_dir = Path(\"Results\")\n",
    "results_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "perf_col = \"aroc\"\n",
    "\n",
    "scaffold_old_col = \"Id\"\n",
    "var_class_col = \"Variant_Classification\"\n",
    "prot_change_col = \"Protein_Change\"\n",
    "gene_name_col = \"gene_name\"\n",
    "\n",
    "perf_delta_col = \"Performance_delta\"\n",
    "scaffold_col = \"scaffold\"\n",
    "\n",
    "promisc_new_col = \"Promiscuity\"\n",
    "bin_id_col = \"Bin_idx\"\n",
    "\n",
    "list_separator = \";\"\n",
    "\n",
    "# Unique colors for variants\n",
    "variants = [\n",
    "    \"Missense_Mutation\",\n",
    "    \"Intron\",\n",
    "    \"Silent\",\n",
    "    \"Splice_Site\",\n",
    "    \"Nonsense_Mutation\",\n",
    "    \"Frame_Shift_Del\",\n",
    "    \"IGR\",\n",
    "    \"3'UTR\",\n",
    "    \"5'Flank\",\n",
    "    \"5'UTR\",\n",
    "    \"RNA\",\n",
    "    \"Frame_Shift_Ins\",\n",
    "    \"In_Frame_Del\",\n",
    "    \"lincRNA\",\n",
    "    \"In_Frame_Ins\",\n",
    "    \"Start_Codon_SNP\",\n",
    "    \"Nonstop_Mutation\",\n",
    "    \"De_novo_Start_OutOfFrame\",\n",
    "    \"De_novo_Start_InFrame\",\n",
    "    \"Stop_Codon_Del\",\n",
    "    \"Start_Codon_Del\",\n",
    "    \"Start_Codon_Ins\",\n",
    "]\n",
    "variants_n = len(variants)\n",
    "variants_colors = {}\n",
    "if variants_n <= 10:\n",
    "    c = {variants[i]: plt.color_sequences[\"tab10\"][i] for i in range(variants_n)}\n",
    "    variants_colors.update(c)\n",
    "else:\n",
    "    c = {variants[i]: plt.color_sequences[\"tab10\"][i] for i in range(10)}\n",
    "    variants_colors.update(c)\n",
    "    vn_rest = variants_n - 10\n",
    "    if vn_rest <= 10:\n",
    "        c = {\n",
    "            variants[i]: plt.color_sequences[\"tab20\"][j]\n",
    "            for i, j in zip(range(10, 10 + vn_rest), range(1, 2 * vn_rest, 2))\n",
    "        }\n",
    "        variants_colors.update(c)\n",
    "    else:\n",
    "        c = {\n",
    "            variants[i]: plt.color_sequences[\"tab20\"][j]\n",
    "            for i, j in zip(range(10, 20), range(1, 20, 2))\n",
    "        }\n",
    "        variants_colors.update(c)\n",
    "        vn_rest -= 10\n",
    "        if vn_rest <= 9:\n",
    "            c = {\n",
    "                variants[i]: plt.color_sequences[\"Pastel1\"][j]\n",
    "                for i, j in zip(range(20, 20 + vn_rest), range(vn_rest))\n",
    "            }\n",
    "            variants_colors.update(c)\n",
    "        else:\n",
    "            raise ValueError(\"No colors left, yet\")\n",
    "# Add color for `Other`\n",
    "variants_colors_other = \"0.7\"\n",
    "\n",
    "# Matplotlib config\n",
    "\n",
    "plt.rcdefaults()\n",
    "\n",
    "main_font_size = 6\n",
    "label_font_size = 10\n",
    "\n",
    "axis_color = \"0.15\"\n",
    "\n",
    "plt.rcParams[\"text.color\"] = axis_color\n",
    "plt.rcParams[\"axes.labelcolor\"] = axis_color\n",
    "plt.rcParams[\"xtick.color\"] = axis_color\n",
    "plt.rcParams[\"ytick.color\"] = axis_color\n",
    "plt.rcParams[\"axes.edgecolor\"] = axis_color\n",
    "\n",
    "plt.rcParams[\"font.size\"] = main_font_size\n",
    "plt.rcParams[\"axes.labelsize\"] = main_font_size\n",
    "plt.rcParams[\"axes.titlesize\"] = main_font_size\n",
    "plt.rcParams[\"xtick.labelsize\"] = main_font_size\n",
    "plt.rcParams[\"ytick.labelsize\"] = main_font_size\n",
    "plt.rcParams[\"legend.fontsize\"] = main_font_size\n",
    "plt.rcParams[\"figure.labelsize\"] = main_font_size\n",
    "plt.rcParams[\"figure.titlesize\"] = main_font_size\n",
    "\n",
    "plt.rcParams[\"axes.spines.top\"] = False\n",
    "plt.rcParams[\"axes.spines.right\"] = False\n",
    "\n",
    "plt.rcParams[\"savefig.transparent\"] = True\n",
    "\n",
    "plt.rcParams[\"pdf.fonttype\"] = 42\n",
    "plt.rcParams[\"ps.fonttype\"] = 42\n",
    "plt.rcParams[\"svg.fonttype\"] = \"none\"\n",
    "\n",
    "#\n",
    "\n",
    "\n",
    "def figsize_in_mm(width, height):\n",
    "    mm = 1 / 25.4  # millimeters in inches\n",
    "    return (width * mm, height * mm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scaffold level\n",
    "\n",
    "# Config\n",
    "config = cfg.configurations[3]\n",
    "mut_remove_results_file = (\n",
    "    results_dir / \"TEST-mutation-remove-experiments-Scaffold_level.csv\"\n",
    ")\n",
    "mut_join_col = scaffold_col\n",
    "promisc_col = \"popcov_but_sqrt\"\n",
    "pd_int_cols = {\"unique_peptides\": \"Int64\"}\n",
    "unique_cols = [\n",
    "    \"Chromosome\",\n",
    "    \"Start_position\",\n",
    "    \"End_position\",\n",
    "    \"unique_peptides\",\n",
    "    \"popcov_but_sqrt\",\n",
    "]\n",
    "best_genes_cols = [\n",
    "    gene_name_col,\n",
    "    prot_change_col,\n",
    "    scaffold_col,\n",
    "    promisc_col,\n",
    "    perf_delta_col,\n",
    "]\n",
    "prot_change_extra_col = prot_change_col + \"_extra\"\n",
    "#\n",
    "\n",
    "model_name = config[\"plot_label\"]\n",
    "\n",
    "mutset = pd.read_csv(\n",
    "    config[\"contig_file\"], sep=\"\\t\", low_memory=False, dtype=pd_int_cols\n",
    ")\n",
    "# Filter the file like in 'DataSets_validation.py' first:\n",
    "if config[\"hotspots\"]:\n",
    "    mutset = mutset.loc[mutset[\"contig\"].notna()]\n",
    "mutset = mutset.dropna(\n",
    "    axis=\"index\", how=\"any\", subset=unique_cols, ignore_index=True\n",
    ").drop_duplicates(unique_cols, ignore_index=True)\n",
    "mutset = pl.DataFrame(mutset)\n",
    "# Renaming for scaffolds only:\n",
    "mutset = mutset.rename({scaffold_old_col: scaffold_col})\n",
    "\n",
    "contigmut = pl.read_csv(mut_remove_results_file)\n",
    "\n",
    "aroccontigmut = contigmut[perf_col][0]\n",
    "\n",
    "contig_mutations_with_delta = (\n",
    "    contigmut.slice(1)\n",
    "    .with_columns((aroccontigmut - pl.col(perf_col)).alias(perf_delta_col))\n",
    "    .sort(perf_delta_col, descending=True)\n",
    ")\n",
    "\n",
    "# Add model performance info to mutations\n",
    "#  The `filter` construction might be scaffold/contig-only\n",
    "mutset = (\n",
    "    mutset.join(\n",
    "        contig_mutations_with_delta,\n",
    "        on=mut_join_col,\n",
    "        suffix=\"_dup\",\n",
    "        how=\"left\",\n",
    "        coalesce=False,\n",
    "    )\n",
    "    .filter(pl.col(mut_join_col).is_not_null())\n",
    "    .sort(perf_delta_col, descending=True)\n",
    ")\n",
    "\n",
    "# Deal with some list columns (scaffold/contig-only)\n",
    "contig_mutations_with_delta = contig_mutations_with_delta.with_columns(\n",
    "    pl.col(prot_change_col).str.split(list_separator),\n",
    "    pl.col(gene_name_col).str.split(list_separator),\n",
    ")\n",
    "contig_mutations_with_delta = contig_mutations_with_delta.with_columns(\n",
    "    pl.col(prot_change_col)\n",
    "    .list.unique()\n",
    "    .list.set_difference([\"\"])\n",
    "    .alias(prot_change_extra_col)\n",
    ").with_columns(\n",
    "    pl.when(pl.col(prot_change_extra_col).list.len() > 0).then(\n",
    "        pl.col(prot_change_extra_col)\n",
    "    )\n",
    ")\n",
    "\n",
    "print(\n",
    "    \"Performance delta limits: (\"\n",
    "    f\"{contig_mutations_with_delta[perf_delta_col].min()}\"\n",
    "    \", \"\n",
    "    f\"{contig_mutations_with_delta[perf_delta_col].max()}\"\n",
    "    \")\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutset = mutset.filter(pl.col(perf_delta_col).is_not_null())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "contig_mutations_with_delta[gene_name_col].str.find(\";;\").is_not_null().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Performance delta distribution\n",
    "\n",
    "# Config\n",
    "promisc_boundary = 13\n",
    "perf_delta_x_range = (0.0, 0.32)\n",
    "\n",
    "fig_size = figsize_in_mm(94, 72)\n",
    "\n",
    "top_genes_n = 100\n",
    "#\n",
    "\n",
    "base_df = contig_mutations_with_delta\n",
    "prot_change_test_col = prot_change_extra_col\n",
    "\n",
    "canonical = base_df.filter(pl.col(prot_change_test_col).is_not_null())\n",
    "non_canonical = base_df.filter(pl.col(prot_change_test_col).is_null())\n",
    "\n",
    "canonical_high = canonical.filter(pl.col(promisc_col) > promisc_boundary)\n",
    "canonical_low = canonical.filter(pl.col(promisc_col) <= promisc_boundary)\n",
    "non_canonical_high = non_canonical.filter(pl.col(promisc_col) > promisc_boundary)\n",
    "non_canonical_low = non_canonical.filter(pl.col(promisc_col) <= promisc_boundary)\n",
    "\n",
    "data_subgroups = {\n",
    "    \"Canonical (High Promiscuity)\": canonical_high,\n",
    "    \"Canonical (Low Promiscuity)\": canonical_low,\n",
    "    \"Non-Canonical (High Promiscuity)\": non_canonical_high,\n",
    "    \"Non-Canonical (Low Promiscuity)\": non_canonical_low,\n",
    "}\n",
    "\n",
    "# Fig\n",
    "\n",
    "fig, axes = plt.subplots(2, 2, figsize=fig_size, layout=\"constrained\")\n",
    "for ax, (title, series) in zip(axes.flatten(), data_subgroups.items()):\n",
    "    h_out = ax.hist(\n",
    "        series[perf_delta_col], bins=20, density=False, range=perf_delta_x_range\n",
    "    )\n",
    "    ax.set_title(title)\n",
    "    ax.set_xlabel(\"Performance Delta\")\n",
    "    ax.set_ylabel(\"Frequency\")\n",
    "    ax.set_yticks([])\n",
    "    ax.set_yticklabels([])\n",
    "    ax.grid(axis=\"y\", alpha=0.5)\n",
    "\n",
    "    # Save top N genes from each bin:\n",
    "\n",
    "    bin_n = h_out[0][::-1]\n",
    "    bin_starts = h_out[1]\n",
    "    bin_ranges = [\n",
    "        (bin_starts[i], bin_starts[i + 1]) for i in range(len(bin_starts) - 1)\n",
    "    ][::-1]\n",
    "    bin_ids = [i for i in range(len(bin_n))][::-1]\n",
    "\n",
    "    top_genes = []\n",
    "\n",
    "    for b_n, (b_s, b_e), b_id in zip(bin_n, bin_ranges, bin_ids):\n",
    "        if b_n == 0:\n",
    "            continue\n",
    "        b_series = series.filter(\n",
    "            (pl.col(perf_delta_col) >= b_s) & (pl.col(perf_delta_col) < b_e)\n",
    "        ).select(pl.col(best_genes_cols))\n",
    "        # Might work only with scaffolds/contigs\n",
    "        b_series = (\n",
    "            b_series.explode([gene_name_col, prot_change_col])\n",
    "            .unique()\n",
    "            .sort(\n",
    "                [promisc_col, gene_name_col, prot_change_col, mut_join_col],\n",
    "                descending=[True, False, False, False],\n",
    "            )\n",
    "            .with_columns(pl.lit(b_id).alias(bin_id_col))\n",
    "        )\n",
    "        #\n",
    "        # Take first top N promiscuity values (dense row selection)\n",
    "        b_series = b_series.with_columns(\n",
    "            pl.col(promisc_col).rank(method=\"dense\", descending=True).alias(\"_rank\")\n",
    "        )\n",
    "        b_series = b_series.filter(\n",
    "            pl.col(promisc_col).rank(method=\"dense\", descending=True) <= top_genes_n\n",
    "        )\n",
    "        #\n",
    "        top_genes.append(b_series)\n",
    "\n",
    "    top_genes = pl.concat(top_genes).sort(\n",
    "        [promisc_col, perf_delta_col, gene_name_col, prot_change_col, mut_join_col],\n",
    "        descending=[True, True, False, False, False],\n",
    "    )\n",
    "    top_genes = top_genes.rename({promisc_col: promisc_new_col})\n",
    "    print(top_genes)\n",
    "\n",
    "    title_mod = title.lower().replace(\"(\", \"\").replace(\")\", \"\").replace(\" \", \"_\")\n",
    "    f_name = f\"top-{top_genes_n}-genes-{title_mod}-{model_name.replace(' ', '_')}.csv\"\n",
    "    # top_genes.write_csv(f_name)\n",
    "for fmt in (\"pdf\", \"png\"):\n",
    "    fig.savefig(\n",
    "        results_dir\n",
    "        / f\"performance-delta-histograms-{model_name.replace(' ', '_')}.{fmt}\",\n",
    "        dpi=600,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variant classifications distribution\n",
    "\n",
    "# Config\n",
    "base_df = mutset\n",
    "prot_change_test_col = prot_change_col\n",
    "\n",
    "fig_size = figsize_in_mm(70.3, 154.3)\n",
    "# n_other_groups = [5, 9]\n",
    "n_other_groups = [3, 0]\n",
    "#\n",
    "\n",
    "canonical = base_df.filter(pl.col(prot_change_test_col).is_not_null())\n",
    "non_canonical = base_df.filter(pl.col(prot_change_test_col).is_null())\n",
    "\n",
    "groups = [\n",
    "    (\"Canonical Mutations\", canonical),\n",
    "    (\"Non-Canonical Mutations\", non_canonical),\n",
    "]\n",
    "\n",
    "# Fig\n",
    "\n",
    "fig, axes_pie = plt.subplots(2, 1, figsize=fig_size, layout=\"constrained\")\n",
    "for ax, (title, group_df), n_other in zip(axes_pie.flatten(), groups, n_other_groups):\n",
    "    counts = group_df.group_by(var_class_col).len().sort(\"len\", descending=True)\n",
    "    labels = counts[var_class_col]\n",
    "    sizes = counts[\"len\"]\n",
    "    n_labels = len(labels)\n",
    "    if 0 < n_other < n_labels:\n",
    "        other_sum = sizes.slice(-n_other).sum()\n",
    "        sizes = sizes.slice(0, n_labels - n_other).append(\n",
    "            pl.Series(\"a\", [other_sum], dtype=pl.UInt32)\n",
    "        )\n",
    "        other_label = \"Other: \" + \",\\n\".join(labels.slice(-n_other).to_list())\n",
    "        labels = labels.slice(0, n_labels - n_other).append(\n",
    "            pl.Series(\"a\", [other_label])\n",
    "        )\n",
    "        n_labels = len(labels)\n",
    "    elif n_other >= n_labels:\n",
    "        raise ValueError(\"Too much labels compressed into `n_other_groups`\")\n",
    "    colors = [\n",
    "        variants_colors_other if lab.startswith(\"Other: \") else variants_colors[lab]\n",
    "        for lab in labels\n",
    "    ]\n",
    "    wedges, _ = ax.pie(\n",
    "        sizes,\n",
    "        labels=labels,\n",
    "        colors=colors,\n",
    "        autopct=None,\n",
    "        startangle=90,\n",
    "        labeldistance=None,\n",
    "    )\n",
    "    ax.set_title(title)\n",
    "    percents = (sizes / sizes.sum() * 100).round(2)\n",
    "    legend_labels = [f\"{lab}--{p}%\" for lab, p in zip(labels, percents)]\n",
    "    ax.legend(wedges, legend_labels, loc=\"center left\", bbox_to_anchor=(1, 0, 0.5, 1))\n",
    "    ax.axis(\"equal\")\n",
    "for fmt in (\"pdf\", \"png\"):\n",
    "    fig.savefig(\n",
    "        results_dir\n",
    "        / f\"variant-classifications-pie-charts-{model_name.replace(' ', '_')}.{fmt}\",\n",
    "        dpi=600,\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "svn-opti",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
